### PR TITLE
[dockerng] Add support of `labels` parameter for dockerng

### PR DESCRIPTION
**Pull request by ticosax**
@terminalmage Can you have a look please ?

I'm not sure about the yaml format.

### CODE REVIEW

File: salt/modules/dockerng.py [hunk 1]
```diff
@@ -1634,6 +1637,26 @@ def _valid_pid_mode():  # pylint: disable=unused-variable
...
+            return
+        try:
+            _valid_stringlist('labels')
+        except SaltInvocationError:
+            try:
+                _valid_dict('labels')
+            except SaltInvocationError:
+                raise SaltInvocationError(
+                    'labels can only be a list of strings or a dict containing'
+                    ' strings')
```
Comment (blueyed) [inline salt/modules/dockerng.py]: Can there be more context in the error message, e.g. a `dockerng:` prefix and/or the name of the state where it's used?  Or is that handled by Salt automatically already?

File: salt/modules/dockerng.py [hunk 2]
```diff
@@ -416,6 +416,9 @@
     'cpuset': {
         'path': 'Config:Cpuset',
     },
+    'labels': {
+      'path': 'Config.Labels',
```
Comment (terminalmage) [inline salt/modules/dockerng.py]: Is ``Labels`` a sub-key of ``Config``? If so, then this should be ``Config:Labels`` (replace the dot with a colon).

File: salt/modules/dockerng.py (hunk 1 above)
Comment (terminalmage) [inline salt/modules/dockerng.py]: Yeah, a prefix would help explain that this is dockerng, this is probably something we can do a better job of in general in this module.

Comment (terminalmage) [inline salt/modules/dockerng.py]: @blueyed You can also get the module name by modifying the log format string.

File: salt/modules/dockerng.py [hunk 3]
```diff
@@ -2630,6 +2653,10 @@ def create(image,
 
             This is only used if Salt needs to pull the requested image.
 
+    labels
+        Add Metadata to the container. Can be a list of strings or a dictionary
+        of strings (keys and values).
+
```
Comment (terminalmage) [inline salt/modules/dockerng.py]: Need example CLI usage like the rest of the parameters.

**Comment by terminalmage** [comment]
The SLS actually looks fine, though I've made a couple other comments about things that need to be updated/fixed.

I'm not a big fan of how PyYAML handles nested dictionaries, to be honest. The double-indentation causes a lot of people problems. I wrote a function called [salt.utils.repack_dictlist()](https://github.com/saltstack/salt/blob/e697326/salt/utils/__init__.py#L2499-L2528) which will convert a list of one-element dicts into a single dictionary, which would allow the SLS to look like this:

```yaml
foo:
  dockerng.running:
    - image: bar/baz:latest
    - labels:
      - KEY1: VALUE1
      - KEY2: VALUE2
```

[salt.utils.repack_dictlist()](https://github.com/saltstack/salt/blob/e697326/salt/utils/__init__.py#L2499-L2528) could be used in ``_valid_labels()`` to repack a list of dicts into a single dicts. The catch here is that if the labels are passed as list of strings then we don't want to use this function. However, using it would allow either the double-indentation (as in your example) and my example above to be used. Using my example SLS is less error-prone for people not familiar with the need to double-indent nested dictionaries.

Thoughts?

File: salt/modules/dockerng.py (hunk 2 above)
Comment (ticosax) [inline salt/modules/dockerng.py]: nice catch !

File: salt/modules/dockerng.py (hunk 1 above)
Comment (ticosax) [inline salt/modules/dockerng.py]: Shall we address that in a separate PR ?

File: salt/modules/dockerng.py (hunk 3 above)
Comment (ticosax) [inline salt/modules/dockerng.py]: ok

**Comment by ticosax** [comment]
@terminalmage @blueyed thank you for your comments.
They have been addressed.
If you :+1: don't merge until I fixup the commits.

**Comment by blueyed** [comment]
LGTM :+1:

**Comment by ticosax** [comment]
Thx @blueyed

**Comment by rallytime** [comment]
@ticosax Looks like you have a couple of [pylint errors](https://jenkins.saltstack.com/job/salt-pr-lint-n/9694/violations/file/salt/modules/dockerng.py/). Would you mind grabbing those as well?

**Comment by ticosax** [comment]
@rallytime Thank you.
I rebased the branch.

**Comment by terminalmage** [comment]
Looks great, thanks!

**Pull request merged**